plutus-core/plutus-ir/src/PlutusIR/Transform/NewLetFloat.hs

```diff
@@ -0,0 +1,435 @@
+-- EXCEPT if the nested let is intercepted by a l/Lambda (depends on a lambda of the parent-let's rhs)
+-- e.g. ok: let {x = \ z -> let {y = z+1} in y} in ...
+mergeBindings :: NE.NonEmpty (LetNaked tyname name uni fun a)
+              -> NE.NonEmpty (Binding tyname name uni fun a)
+mergeBindings = foldMap1 (^.letNBs)
+    -- we ignore annotations and recursivity, the letholes *must be* merged together with a Recursive let bindings
+    -- the order of (<>) does not matter because it is a recursive let-group anyway.
+
+mergeLetsIn :: NE.NonEmpty (LetNaked tyname name uni fun a) -> Term tyname name uni fun a -> Term tyname name uni fun a
+mergeLetsIn ls =
+                 -- TODO: use some semigroup annotation-joining
+                 -- arbitrary: use the annotation of the first let of the floated lets as the annotation of the new let
+                 Let (NE.head ls^.letNAnn)
+                 Rec -- needs to be rec because we don't do dependency resolution at this pass, See Note [LetRec splitting pass]
+                 (mergeBindings ls)
+
+
+floatTerm :: (PLC.ToBuiltinMeaning uni fun,
+            PLC.HasUnique tyname PLC.TypeUnique, PLC.HasUnique name PLC.TermUnique,
+            Ord tyname, Ord name
+            )
+          => Term tyname name uni fun a -> Term tyname name uni fun a
+floatTerm t =
+    let ms = mark t
+        (letholes, t') = removeLets ms t
+    in floatBackLets letholes t'
+
+-- CONSTANTS
+
+-- for simplicity, the top position is also linked to a unique number
+-- chosen to not clash with any actual uniques of names/tynames of the program
+topUnique :: PLC.Unique
+topUnique = coerce (-1 :: Int)
+
+-- arbitrarily chosen
+topDepth :: Depth
+topDepth = -1
+
+-- arbitrary chosen as LetRhs, because top conceptually can be thought as a global let-rhs.
+topType :: PosType
+topType = LetRhs
+
+topPos :: Pos
+topPos = (topDepth, topUnique, topType)
+
+-- HELPERS
+
+maxPos :: M.Map k Pos -> Pos
+maxPos = foldr max topPos . M.elems
+
+withDepth :: (r ~ MarkCtx, MonadReader r m)
+          => (Depth -> Depth) -> m a -> m a
+withDepth = local . over _1
+
+withLam :: (r ~ MarkCtx, MonadReader r m, PLC.HasUnique name unique)
+        => name
+        -> m a -> m a
+withLam n = local $ \ (d, scope) ->
+    let u = n^.PLC.theUnique
+        d' = d+1
+        pos' = (d', u, LamBody)
+    in (d', M.insert u pos' scope)
+
+withBs :: (r ~ MarkCtx, MonadReader r m, PLC.HasUnique name PLC.TermUnique, PLC.HasUnique tyname PLC.TypeUnique)
+       => NE.NonEmpty (Binding tyname name uni fun a3)
+       -> Pos
+       -> m a -> m a
+withBs bs pos = local $ second (M.fromList [(bid,pos) | bid <- bs^..traversed.bindingIds] <>)
+
+
+unmovable :: PLC.ToBuiltinMeaning uni fun => LetProduct tyname name uni fun a -> Bool
+unmovable = \case
+    LetProduct _ Rec bs _ ->  any mayHaveEffects bs
+    LetProduct _ NonRec (b  NE.:|[]) _  -> mayHaveEffects b
+    LetProduct _ NonRec _ _ -> error "should not happen because we prior break down nonrecs"
```

Comment:
Depending how you feel on partial pattern matching, there was an internal discussion, where Duncan was suggesting in favor of partial pattern matching, as partial pattern matching will report the file and line number in case of that error happening.

Especially, as this is the kind of error that we cannot really recover from...

The whole discussion is actually quite informative here:
https://input-output-rnd.slack.com/archives/C48HA39RS/p1628118388025500?thread_ts=1628117570.024400&cid=C48HA39RS